src/Raven.Server/Documents/Sharding/Handlers/IncomingExternalReplicationHandlerForShard.cs

```diff
@@ -0,0 +1,116 @@
+using Raven.Server.Documents.TransactionMerger.Commands;
+using Raven.Server.ServerWide.Context;
+using Raven.Server.Utils;
+using Sparrow.Json;
+
+namespace Raven.Server.Documents.Sharding.Handlers
+{
+    public class IncomingExternalReplicationHandlerForShard : IncomingReplicationHandler
+    {
+        private readonly ShardedDocumentDatabase _shardedDatabase;
+
+        public IncomingExternalReplicationHandlerForShard(
+            TcpConnectionOptions options,
+            ReplicationLatestEtagRequest replicatedLastEtag,
+            ReplicationLoader parent, JsonOperationContext.MemoryBuffer bufferToCopy,
+            ReplicationLatestEtagRequest.ReplicationType replicationType) : base(options, replicatedLastEtag, parent, bufferToCopy, replicationType)
+        {
+            _shardedDatabase = ShardedDocumentDatabase.CastToShardedDocumentDatabase(parent.Database);
+        }
+
+        protected override DocumentMergedTransactionCommand GetMergeDocumentsCommand(DocumentsOperationContext context,
+            DataForReplicationCommand data, long lastDocumentEtag)
+        {
+            return new MergedDocumentReplicationForShardCommand(_shardedDatabase, data, lastDocumentEtag);
+        }
+
+        internal class MergedDocumentReplicationForShardCommand : MergedDocumentReplicationCommand
+        {
+            private readonly ShardedDocumentDatabase _database;
+
+            public MergedDocumentReplicationForShardCommand(ShardedDocumentDatabase database, DataForReplicationCommand replicationInfo, long lastEtag) : base(replicationInfo, lastEtag)
+            {
+                _database = database;
+            }
+
+            protected override ChangeVector PreProcessItem(DocumentsOperationContext context, ReplicationBatchItem item)
+            {
+                var changeVector = context.GetChangeVector(item.ChangeVector);
+                var dbId = _database.DbBase64Id;
+
+                if ((changeVector.IsSingle && changeVector.Contains(dbId)) ||
+                    changeVector.Order.Contains(dbId) ||
+                    changeVector.Version.Contains(dbId))
+                {
+                    // this item already exists in storage
+                    // was sent directly back from replication (in case of bidirectional replication) or changed in the destination
+                    // either way, since the dbId exists in the item's change vector we can safely return without updating
+                    return changeVector.Order;
+                }
+
+                if (HasConflicts(context, item, changeVector))
+                    return changeVector.Order;
+
+                var shouldUpdateDatabaseCv = true;
+                var current = context.LastDatabaseChangeVector ?? DocumentsStorage.GetDatabaseChangeVector(context);
+
+                if (current.IsNullOrEmpty == false)
+                {
+                    var dbIdsFromDestCv = ChangeVector.ExtractDbIdsFromChangeVector(changeVector.Order);
```

Comment:
I don't understand why does it matter..
if you have replication A (non-sharded) <=> B (sharded)
modifications on B will "leak" the change vector to A, but you still will have the same issue with missing documents for the A -> B replication, no?